packages/qwik/src/core/v2/shared/component-execution.ts

```diff
@@ -104,45 +118,65 @@ export const executeComponent2 = (
  * So when executing a component we only care about its last JSX Output.
  */
 
-function addUseOnEvents(jsx: JSXOutput, useOnEvents: UseOnMap) {
-  let jsxElement = findFirstStringJSX(jsx);
-  let isInvisibleComponent = false;
-  if (!jsxElement) {
-    /**
-     * We did not find any jsx node with a string tag. This means that we should append:
-     *
-     * ```html
-     * <script type="placeholder" hidden on-document:qinit="..."></script>
-     * ```
-     *
-     * This is needed because use on events should have a node to attach them to.
-     */
-    isInvisibleComponent = true;
-    jsxElement = addScriptNodeForInvisibleComponents(jsx);
+function addUseOnEvents(
+  jsx: JSXOutput,
+  useOnEvents: UseOnMap
+): ValueOrPromise<JSXNode<string> | null> {
+  const jsxElement = findFirstStringJSX(jsx);
+  return maybeThen(jsxElement, (jsxElement) => {
+    let isInvisibleComponent = false;
     if (!jsxElement) {
-      return;
+      /**
+       * We did not find any jsx node with a string tag. This means that we should append:
+       *
+       * ```html
+       * <script type="placeholder" hidden on-document:qinit="..."></script>
+       * ```
+       *
+       * This is needed because use on events should have a node to attach them to.
+       */
+      isInvisibleComponent = true;
     }
-  }
-  for (const key in useOnEvents) {
-    if (Object.prototype.hasOwnProperty.call(useOnEvents, key)) {
-      let props = jsxElement.props;
-      if (props === EMPTY_OBJ) {
-        props = jsxElement.props = {};
-      }
-      let propValue = props[key] as UseOnMap['any'] | UseOnMap['any'][0] | undefined;
-      if (propValue === undefined) {
-        propValue = [];
-      } else if (!Array.isArray(propValue)) {
-        propValue = [propValue];
+    for (const key in useOnEvents) {
+      if (Object.prototype.hasOwnProperty.call(useOnEvents, key)) {
+        if (isInvisibleComponent) {
+          if (key === 'onQvisible$') {
+            jsxElement = addScriptNodeForInvisibleComponents(jsx);
+            if (jsxElement) {
+              addUseOnEvent(jsxElement, 'document:onQinit$', useOnEvents[key]);
+            }
+          } else if (isDev) {
+            logWarn(
+              'You are trying to add an event using `useOn` hook, ' +
+                'but a node to which you can add an event is not found. ' +
+                'Please make sure that the component has a valid element node. '
+            );
+          }
```

Comment:
I have added a warning message for dev mode if someone is trying to use `useOn` hook on the invisible component. We can't handle that, because we don't have any element node to add an event to.